### PR TITLE
feat: merkleization of vectors

### LINKED ISSUE #598: [SSZ] Support merkleization for vectors
## Description 

We need to add support for hash_tree_root(vector)

Spec is [here](https://github.com/ethereum/consensus-specs/blob/b2f2102dad0cd8b28a657244e645e0df1c0d246a/ssz/simple-serialize.md#merkleization)

### PR DESCRIPTION
**Pull request by Godspower-Eze**
Closes #598

### CODE REVIEW

**Review by MegaRedHand — APPROVED** [review]
Good work! Left some minor comments, but feel free to merge this as-is.

File: lib/ssz_ex.ex [hunk 1]
```diff
@@ -100,14 +113,23 @@ defmodule LambdaEthereumConsensus.SszEx do
     end
   end
 
+  @spec hash_tree_root_vector_basic_type(binary()) ::
+          {:ok, Types.root()} | {:error, String.t()}
+  def hash_tree_root_vector_basic_type(chunks) do
+    chunks_len = chunks |> get_chunks_len()
+    leaf_count = chunks_len |> next_pow_of_two()
```
[the remark this answers is not in the document]
Comment (Godspower-Eze) [reply lib/ssz_ex.ex:120]: You mean, this?

```
leaf_count = chunks |> get_chunks_len() |> next_pow_of_two()
```

File: lib/ssz_ex.ex [hunk 2]
```diff
@@ -87,10 +87,23 @@ defmodule LambdaEthereumConsensus.SszEx do
...
 
+  @spec hash_tree_root(list(), {:vector, any, non_neg_integer}) ::
+          {:ok, Types.root()} | {:error, String.t()}
+  def hash_tree_root(vector, {:vector, type, size}) do
+    if variable_size?(type) do
+      # TODO
+      # hash_tree_root_vector_complex_type(vector, {:vector, type, size}, limit)
+      {:error, "Not implemented"}
+    else
+      packed_chunks = pack(vector, {:list, type, size})
```
[the remark this answers is not in the document]
Comment (Godspower-Eze) [reply lib/ssz_ex.ex:98]: How would you go about this?

I was thinking of something like this:

```
def pack(list, {:list, type, _size} | {:vector, type, _size} = schema) do
```

but this is not valid syntax

File: lib/ssz_ex.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (MegaRedHand) [reply lib/ssz_ex.ex:120]: Yes!

File: lib/ssz_ex.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (MegaRedHand) [reply lib/ssz_ex.ex:98]: You can use a guard clause:
```elixir
  def pack(vector, {type, _, _}) when type in [:vector, :list] do
```

File: lib/ssz_ex.ex [hunk 3]
```diff
@@ -658,4 +680,8 @@ defmodule LambdaEthereumConsensus.SszEx do
 
     <<left::binary, new_chunk::binary, right::binary>>
   end
+
+  defp get_chunks_len(chunks) do
+    chunks |> byte_size() |> div(@bytes_per_chunk)
+  end
```
[the remark this answers is not in the document]
Comment (Godspower-Eze) [reply lib/ssz_ex.ex:685]: You mean guards that check `chunks_len == 1` and `chunks_len > limit`?

[the remark this answers is not in the document]
Comment (MegaRedHand) [reply lib/ssz_ex.ex:685]: Exactly

[the remark this answers is not in the document]
Comment (Godspower-Eze) [reply lib/ssz_ex.ex:685]: I think it's overkill in this case

[the remark this answers is not in the document]
Comment (MegaRedHand) [reply lib/ssz_ex.ex:685]: Yeah, I thought so 😅. Let's just merge then!

**Review by MegaRedHand — APPROVED** [review]

**Pull request merged**